test/parallel/test-fs-cp.js

```diff
@@ -0,0 +1,766 @@
+// copied over.
+{
+  const src = dirname(require.resolve('../fixtures/copy/kitchen-sink'));
+  const dest = nextdir();
+  cpSync(src, dest, { recursive: true });
+  cp(src, dest, {
+    dereference: true,
+    errorOnExist: true,
+    force: false,
+    recursive: true,
+  }, common.mustCall((err) => {
+    assert.strictEqual(err.code, 'ERR_FS_CP_EEXIST');
+  }));
+}
+
+// It returns EEXIST error if attempt is made to copy symlink over file.
+{
+  const src = nextdir();
+  mkdirSync(join(src, 'a', 'b'), { recursive: true });
+  symlinkSync(join(src, 'a', 'b'), join(src, 'a', 'c'));
+
+  const dest = nextdir();
+  mkdirSync(join(dest, 'a'), { recursive: true });
+  writeFileSync(join(dest, 'a', 'c'), 'hello', 'utf8');
+  cp(src, dest, { recursive: true }, common.mustCall((err) => {
+    assert.strictEqual(err.code, 'EEXIST');
+  }));
+}
+
+// It makes file writeable when updating timestamp, if not writeable.
+{
+  const src = nextdir();
+  mkdirSync(src, { recursive: true });
+  const dest = nextdir();
+  mkdirSync(dest, { recursive: true });
+  writeFileSync(join(src, 'foo.txt'), 'foo', { mode: 0o444 });
+  cp(src, dest, {
+    preserveTimestamps: true,
+    recursive: true,
+  }, common.mustCall((err) => {
+    assert.strictEqual(err, null);
+    assertDirEquivalent(src, dest);
+    const srcStat = lstatSync(join(src, 'foo.txt'));
+    const destStat = lstatSync(join(dest, 'foo.txt'));
+    assert.strictEqual(srcStat.mtime.getTime(), destStat.mtime.getTime());
+  }));
+}
+
+// It copies link if it does not point to folder in src.
+{
+  const src = nextdir();
+  mkdirSync(join(src, 'a', 'b'), { recursive: true });
+  symlinkSync(src, join(src, 'a', 'c'));
+  const dest = nextdir();
+  mkdirSync(join(dest, 'a'), { recursive: true });
+  symlinkSync(dest, join(dest, 'a', 'c'));
+  cp(src, dest, { recursive: true }, common.mustCall((err) => {
+    assert.strictEqual(err, null);
+    const link = readlinkSync(join(dest, 'a', 'c'));
+    assert.strictEqual(link, src);
+  }));
+}
+
+// It accepts file URL as src and dest.
+{
+  const src = dirname(require.resolve('../fixtures/copy/kitchen-sink'));
+  const dest = nextdir();
+  cp(pathToFileURL(src), pathToFileURL(dest), { recursive: true },
+     common.mustCall((err) => {
+       assert.strictEqual(err, null);
+       assertDirEquivalent(src, dest);
+     }));
+}
+
+// It throws if options is not object.
+{
+  assert.throws(
+    () => cp('a', 'b', 'hello', () => {}),
+    { code: 'ERR_INVALID_ARG_TYPE' }
+  );
+}
+
+// Promises implementation of copy.
+
+// It copies a nested folder structure with files, folders, symlinks.
+{
+  const src = dirname(require.resolve('../fixtures/copy/kitchen-sink'));
+  const dest = nextdir();
+  async function testCase() {
+    const p = await fs.promises.cp(src, dest, { recursive: true });
+    assert.strictEqual(p, undefined);
+    assertDirEquivalent(src, dest);
+  }
+  testCase();
+}
```

Comment:
can be done in a separate PR, but if the file is renamed with the `.mjs` extension, these could use top-level await

Reply:
On this point, we should ensure that the resulting promise settles before the end of the process.
```suggestion
// It copies a nested folder structure with files, folders, symlinks.
{
  const src = dirname(require.resolve('../fixtures/copy/kitchen-sink'));
  const dest = nextdir();
  async function testCase() {
    const p = await fs.promises.cp(src, dest, { recursive: true });
    assert.strictEqual(p, undefined);
    assertDirEquivalent(src, dest);
  }
  testCase().then(common.mustCall());
}
```

Reply:
I'm going to go ahead and convert to `.mjs`.